### PR TITLE
Password hashing, credentialsDAO and separate Authenticator

### PR DESCRIPTION
**Pull request by AleksTeresh**
Resolved Issue #14

### CODE REVIEW

File: src/main/java/fi/istrange/traveler/resources/AuthResource.java [hunk 1]
```diff
@@ -36,21 +37,8 @@ public DefaultJwtCookiePrincipal login(
...
-                .fetchOptional();
-
-        if (
-                !result.isPresent() ||
-                !userCredentialsView.getPassword().equals(result.get().getValue(Tables.USER_CREDENTIALS.PASSWORD)) ||
-                !result.get().getValue(Tables.USER_CREDENTIALS.ACTIVE)
-        ) {
-            throw new NotAuthorizedException("Invalid credentials");
+        if (!auth.authenticate(userCredentialsView.getName(), userCredentialsView.getPassword(), database)) {
+            return null;
```
Comment (kujiraOo) [inline src/main/java/fi/istrange/traveler/resources/AuthResource.java]: Wouldn't it be better to throw the NotAuthorized exception here?

File: src/main/java/fi/istrange/traveler/auth/Authenticator.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+
+        Optional<UserCredentials> credentials= credentialDao.fetchByUsername(username, database);
+
+        if (
+                !credentials.isPresent() ||
+                !hashedPassword.equals(credentials.get().getPassword()) ||
+                !credentials.get().getActive()
+        ) {
+            throw new NotAuthorizedException("Invalid credentials");
```
Comment (kujiraOo) [inline src/main/java/fi/istrange/traveler/auth/Authenticator.java]: Wouldn't it be better just to return the result of the check?

File: src/main/java/fi/istrange/traveler/resources/AuthResource.java (hunk 1 above)
Comment (AleksTeresh) [reply src/main/java/fi/istrange/traveler/resources/AuthResource.java]: I wanted that notAuth exceptions are theoiwn from Authorizer only

File: src/main/java/fi/istrange/traveler/auth/Authenticator.java (hunk 2 above)
Comment (AleksTeresh) [reply src/main/java/fi/istrange/traveler/auth/Authenticator.java]: Makes sense. I was thinking about keeping exceptions as they are, but yeah. I'll chnage it in a moment

File: src/main/java/fi/istrange/traveler/resources/AuthResource.java (hunk 1 above)
Comment (kujiraOo) [reply src/main/java/fi/istrange/traveler/resources/AuthResource.java]: I missed the fact, that the authenticate can throw the 422 exception.

Maybe it's better to add the throws signature to authenticate. And then propagate the exceptions to the resource and handle/throw it there. As I see this conditional statement will never be executed and returning the null value, that will never be returned, is confusing a bit.

Comment (AleksTeresh) [reply src/main/java/fi/istrange/traveler/resources/AuthResource.java]: True

File: src/main/java/fi/istrange/traveler/resources/AuthResource.java [hunk 3]
```diff
@@ -36,22 +37,7 @@ public DefaultJwtCookiePrincipal login(
...
-                .fetchOptional();
-
-        if (
-                !result.isPresent() ||
-                !userCredentialsView.getPassword().equals(result.get().getValue(Tables.USER_CREDENTIALS.PASSWORD)) ||
-                !result.get().getValue(Tables.USER_CREDENTIALS.ACTIVE)
-        ) {
-            throw new NotAuthorizedException("Invalid credentials");
-        }
+        auth.authenticate(userCredentialsView.getName(), userCredentialsView.getPassword(), database);
```
Comment (kujiraOo) [inline src/main/java/fi/istrange/traveler/resources/AuthResource.java:40]: It seems to me that not authorized exception if not being thrown now at all. And the client gets the cookie no matter what the result of the check is.

Comment (AleksTeresh) [reply src/main/java/fi/istrange/traveler/resources/AuthResource.java:40]: Just tested it, the exception is thrown and a client gets a response with it

Comment (kujiraOo) [reply src/main/java/fi/istrange/traveler/resources/AuthResource.java:40]: I pulled the PR and tested it, the authentication behaves the way I expected. As long as the body of the request is ok, it gives me a cookie, not matter what the values for the password and username are.

Comment (AleksTeresh) [reply src/main/java/fi/istrange/traveler/resources/AuthResource.java:40]: Sorry, my bad. I was testing violataion of the body structure

**Pull request merged**